### ISSUE TITLE
Memory leak in forge vm.snapshot and vm.revertTo

### ISSUE THREAD

**Issue by maa105** [foundry-rs/foundry#6411]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (7b45265 2023-11-21T00:22:00.899582254Z)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

We have a huge test case where we have 1000s of cases and each with lots of vm.snapshot and vm.revertTo calls. We noticed the memory consumption increases when time goes so we did a small test to confirm it

```sol
contract SnapshotTest is Test {
    function testSnapshots() public {
        uint preState;
        for(uint c = 0; c < 1000; c++) {
            for(uint cc = 0; cc < 10000; cc++) {
                preState = vm.snapshot();
                vm.revertTo(preState);
            }
            vm.sleep(2500);
        }
    }
}
```

What this code does is create a snapshot and revert directly in a loop then sleep for 2.5 seconds. If you run this you'll see the memory usage increases every 2-3 seconds consistently.

While digging around, we noticed that in your documentation [here](https://book.getfoundry.sh/cheatcodes/snapshots#description) you mentioned that reverting a snapshot deletes the snapshot and all subsequent ones. However, we also noticed that [this PR](https://github.com/foundry-rs/foundry/pull/5487) seems to do the opposite. We suspect that the memory leak is related to this PR.

On the short term is there a way to delete/clean the snapshots?

**Comment by mattsse** [comment]
thanks for this, checking shortly.

> On the short term is there a way to delete/clean the snapshots?

there's not but we can certainly add a new cheatcode for that. do you want to try this?

**Comment by mattsse** [comment]
oh I think I remember why we did this, it was just more convenient than the old behaviour but we def need a cheatcode to clean snapshots then

cc @Evalir 

@maa105 
cheatcodes are defined like this

https://github.com/foundry-rs/foundry/blob/388e181c50c9313ca8cfe2c4f413397100dd1ce7/crates/cheatcodes/spec/src/vm.rs#L412-L416

so I think we want a new cheatcode `clearSnapshots()`

and an implementation like revertTo that clears them:

https://github.com/foundry-rs/foundry/blob/388e181c50c9313ca8cfe2c4f413397100dd1ce7/crates/cheatcodes/src/evm.rs#L327-L341

perhaps even an additional cheatcode `RevertToAndRemove` or something that reverts _and_ deletes

do you feel comfortable with giving this a try?

**Comment by maa105** [comment]
Thanks. Ill give it a shot tomorrow

**Comment by maa105** [comment]
Reverting the commit of https://github.com/foundry-rs/foundry/pull/5487 improved the memory leak meaning the increase in memory over time decreased however there is still memory leak :/

**Comment by Evalir** [comment]
hey @maa105 — yep, reverting #5487 should alleviate the problem, but this is not something we want to do—we have users who depend on the new behavior (and we should update the docs) as it's much more convenient to not have snapshots deleted unless specifically requested. The solution to this is what @mattsse proposed

**Comment by maa105** [comment]
The PR is indeed very useful however it worsens the problem of memory leak. Is there a way to reset **all** memory usages inside the test. I.e. mid way inside a test id like to nuke all memory and return to the state when the test started.

**Comment by mattsse** [comment]
https://github.com/foundry-rs/foundry/issues/6411#issuecomment-1824576950

**Comment by Evalir** [comment]
Closed by #6548
